### DISCUSSION TITLE
Setting up some kind of management,... how to start with open source

### DISCUSSION THREAD

**Discussion by WaldemarH** [OpenSalamander/salamander#11] [General]
Hi guys.

I see that there is already something going on which is wunderful. :)

1. The idea to translate comments from czech to english is in my view proper first step.

2. [take all this with a grain of salt as I could be full of bull and OS needs this structure]
Next I was thinking to do some cleaning up as having 90 projects in a solution is not a joyful experience. :)
The issue is that if we start working on this and people will pull the code in various directions it will soon become a mess.

I don't know how the code is structured but the following info gives us some rough estimates of number of files/loc/...

    ---------------------------------------------------------------------------------------
    Language                             files          blank        comment           code
    ---------------------------------------------------------------------------------------
    C++                                    914          71261          58017         578740
    C                                      217          27936          83760         231023
    C/C++ Header                          1111          45755          66176         188867
    INI                                     22            476          25414          23916
    HTML                                   314           4652              0          19310
    Windows Resource File                  200           3325           3044          18972
    Pascal                                  23           2048           1866          12998
    DOS Batch                               51            343             28           1935
    C#                                      16            342           1006           1687
    IDL                                      1            168              0           1502
    JavaScript                              33            207            446           1462
    PHP                                      3            154             23           1248
    CSS                                      4            231              8           1070
    XML                                     16              0              8            676
    WiX source                               1             52              2            342
    MSBuild script                           2              0              7            296
    CMake                                    2             60             16            284
    make                                     7             17             10            247
    Ruby                                     7             27             40            175
    Windows Module Definition               44             48             21            173
    Markdown                                 1             55              0            151
    Visual Basic                            11             42             48            122
    Lua                                      1              8             16             24
    ---------------------------------------------------------------------------------------
    SUM:                                  3001         157207         239956        1085220
    ---------------------------------------------------------------------------------------

Now I don't know how the code is stuctured but in my eyes I would separate salamander,
plugins, help,... all into their own folder and structure. And also have a SDK folder
where you would define all the externals.

Example:

    OS/
      _build/
            debug/
                 x32/
                 x64/
            release/
                   x32/
                   x64/
            populate.cmd
                   
      help/
         plugins/
              'pluginName'/
                          sources/    (html probably)
                          resources/  (images,....)
              ...
              
         salamander/
                   source/
                   resources/
    
         build.cmd  
    
      images/   (allow artists to work on icons,... in their own folders)
            imageName/
                     what ever structure
    
      licenses/
              (all the licenses of external libraries, dlls,...)
              
      plugins/
             'pluginName'/
                         sources/
                         resources/
                         translations/
             ...
             solutionFiles
             
      salamander/
                source/
                      subDirectories (similarly how C# does stuff -> if there is a folder in project it is on disk too)
                resource/
                translations/
                solutionFile
      sdk/
         hhw (Html Help Workshop)
         ... (which ever tools that are needed which are not part of VS by default)
    
      test/   (intended in case we would add the support for gui testing)
      

In this way everything is separated and people would be able to clone, do the minimal to setup sdk folder and start working. At least to me lots of open source projects are beyond a mess and you can spend days before you can build the project, because of all the dependencies.

For sdk folder we would prepare a howTo.txt file which would define urls and steps to add the dependencies to the sdk folder. Now all the dependencies that we can include by default we would already provide.. I don't see the reason why would we be able to include hhw already as it's an old/obsolete app anyway (if nothing else at least a zip file)... the M$ page download doesn't work anymore [https://learn.microsoft.com/en-us/previous-versions/windows/desktop/htmlhelp/microsoft-html-help-downloads](url) ).

3. Next thing would be to setup auto builds as already discussed in issues.

4. Create git brances.
   I would create original branch and put the original code there i.e. how it was at the first open sourcing, so that it will be preserved for ever.
   
   Then I would create feature branches and release-x.y branches.
   
5. Also discussion of the rights would also be appropriate... we don't want somebody to add a backdoor or something to the app that is lots of times running in admin mode. :)

Anyway enough for today... give me your thougths. :)

Br
Waldemar

**Comment by janrysavy** [comment]
Just a note on the directory structure. The current state is based on requirements that are no longer valid. Personally, I'm not thrilled with the .vcxproj placement in the vcxproj subdirectory, I'd rather see it one level above the source code.

**Reply by WaldemarH** [reply]
Ok so let me start working on the structure in another thread and also fixing the batch files (they have issues if root folder path contains spaces).

So:
1. the translations (already under way.. thanks Michal, Eric, David and Jan)
2. create a list of people that are contributing and who has some specific role.. if it can be it would be good to have some kind of preliminary meeting where we can talk and discuss the steps, so that it wont become a mess... having things organized, even if it would be as simple as writing something in a txt file (what somebody does) is really important to have a stable development and to know what kind of help (people) do we need.
Also adding people to https://github.com/OpenSalamander would be good in case something happens to you others can take over and continue. And also the work needs to be shared as we all have our jobs and families. :)
3. organize the folder structure,solutions,projects,.. I would really push this one at this time as doing any reorganization/cleanup down the line would be almost impossible
We'll have a temporary unbuildable project, but will help us a lot down the line.
4. update/fix commands (maybe even convert them to powershell scripts.. if needed) 
5. do github automatization (builds,... )
6. write a list of wishes (a simple txt file would be more then enough)
7. take a few smaller wishes, define what should be done and how (design wise), work on them and create 4.0.1 release (so that we'll get the steps down how to do it)
8. ...

For the git stuff my preference is to have people part of OS project and not doing forks and pull requests as in this way it would be easier to know what somebody is doing and directing the work into proper direction.

Would you be willing to have a video chat and maybe even invite others?

BrW

**Reply by janrysavy** [reply]
Can you please elaborate on point 4. and "have people part of OS project and not doing forks and pull requests"? You mean not using the standard pull-request/review workflow?

**Reply by melloware** [reply]
I like all these ideas except the one about not using the normal fork / PR /code review process. That is the whole point and power of GitHub and open source no?

**Reply by WaldemarH** [reply]
Hi guys.

Yeah it's not a problem.. all these are just ideas on how we could create somekind of organized team.
Now I have never worked with the fork environment, but always git (clone) directly with feature threads where they are then merged into the main branch. So in that way we always have a track of all the features/threads that were worked on.

Can you create a feature thread on your fork, request a pull and that the main repo gets fork's thread that can then be merged? .. it's really the same then as having a managed team. My wish was just that we don't merge directly to main branch, but have some kind of mechanism to track what was worked on and what was the exact state of specific feature before the merge.

The whole idea was a more tightly controlled "development/code review" with all the history inside.

And if you say that we'll use fork/pull then it's fork/pull. :)

**Reply by manison** [reply]
> My wish was just that we don't merge directly to main branch, but have some kind of mechanism to track what was worked on and what was the exact state of specific feature before the merge.

There are many workflows that can be used to manage this. The use of [Gitflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow) proved useful in my job.

**Comment by melloware** [comment]
@manison I am not trying to start a holy war but GitFlow is dead and even Atlassian has stopped recommending it.  See: https://georgestocker.com/2020/03/04/please-stop-recommending-git-flow/

GitHub handles this already perfectly with an excellent workflow.

**Comment by WaldemarH** [comment]
Hi guys.

I've read and watched about git flows and using github fork flow is ok with me. :) ... I've always worked only with closed organizations, so open source fork is new to me. It really just eliminates the step that developers shold be added to the organization, before they can work on the project.

Now there are still some open points to discuss:
1. to have some additional members at OS organization, just to be future safe in case something happens to Jan
2. should we create feature branches on forks and then do pull requests from them?
3. should we always do rebase before pushing the changes.. to push merge conflicts to the author of the feature?
Also at this point I would also have a minimal organization that would define what is somebody working on so that we would try to stay away from each other and have the least merge conflicts as there can be... those can bring issues into the code and are in 99% easily eliminated if people are not working on the same thing. Does GitHug have something like that?

**Reply by melloware** [reply]
Answers:

1. ???
2. Absolutely this should be done on a fork and PR submitted for review https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests
3. Yes and GitHub even has an automatic feature for merging changes from `main` to your feature branch PR so the build can run if its enabled by the Project Owner. For example:

![image](https://github.com/OpenSalamander/salamander/assets/4399574/2cee7be0-ecfb-4f63-a924-9ae1cfbfd09b)

**Comment by janrysavy** [comment]
>     1. to have some additional members at OS organization, just to be future safe in case something happens to Jan

That's a good point, thank you. There are currently two owners of [OpenSalamander](https://github.com/OpenSalamander): @janrysavy and @petrsol. I sent an invitation to the third co-author of Salamander and our server administrator: @KickerTom.

**Reply by WaldemarH** [reply]
Thanks Jan. :)
